src/_pytest/fixtures.py

```diff
@@ -936,6 +956,44 @@ def _ensure_immutable_ids(ids):
     return tuple(ids)
 
 
+def wrap_function_to_warning_if_called_directly(function, fixture_marker):
+    """Wrap the given fixture function so we can issue warnings about it being called directly, instead of
+    used as an argument in a test function.
+
+    The warning is emitted only in Python 3, because I didn't find a reliable way to make the wrapper function
+    keep the original signature, and we probably will drop Python 2 in Pytest 4 anyway.
+    """
+    if six.PY2:
+        return function
+
+    is_yield_function = is_generator(function)
+    msg = FIXTURE_FUNCTION_CALL.format(name=fixture_marker.name or function.__name__)
+    warning = RemovedInPytest4Warning(msg)
+
+    if is_yield_function:
+
+        @functools.wraps(function)
+        def result(*args, **kwargs):
+            __tracebackhide__ = True
+            __being_called_by_pytest = kwargs.pop("__being_called_by_pytest", False)
```

Comment:
instead of introducing this kwarg, why not just unpack the real function from the definition object

Reply:
this seems to me like we should open a follow-up issue and go for the mechanism you currently have as enabling the cleaner way requires a re-factoring/bugfixing in fixture lookup/requesting

Reply:
Just to make sure I follow, keep using the keyword argument approach, and warning only in Python 3? I can try to make it work with Python 2 if you think setting `__wrapped__` will work.

Reply:
please try setting __wrapped__ - wrt the keyword argument approach - i dont like it but as far as i understood we need to sort out a structural thing to get real extraction working

Reply:
so we keep kwargs, and we try to set up the warnings for py2 as well

Reply:
Follow up in #3720